suite/tests/api/ir_x86.c

```diff
@@ -787,6 +787,98 @@ test_instr_encode_and_decode(void *dc, instr_t *instr, uint len_expect,
     instr_destroy(dc, decin);
 }
 
+bool
+instr_same_synthetic(instr_t *instr_orig, instr_t *instr_copy)
+{
+    /* Check that instr_t ISA modes are the same.
+     */
+    if (instr_get_isa_mode(instr_orig) != instr_get_isa_mode(instr_copy))
+        return false;
+
+    /* Check that instr_t categories are the same.
+     */
+    if (instr_get_category(instr_orig) != instr_get_category(instr_copy))
+        return false;
+
+    /* Check that register destination operands are the same.
+     */
+    uint original_num_dsts = (uint)instr_num_dsts(instr_orig);
+    uint num_reg_dsts = 0;
+    for (uint i = 0; i < original_num_dsts; ++i) {
+        opnd_t dst_opnd = instr_get_dst(instr_orig, i);
+        if (opnd_is_reg(dst_opnd)) {
+            // TODO i#6662: we need to compare the virtual register stored in instr_copy
+            // against the converted DR_REG_ (to virtual register) of instr_orig.
+            // Currently we don't have virtual registers, so we can compare the two
+            // DR_REG_ of instr_orig and instr_copy directly.
+            if (!opnd_same(dst_opnd, instr_get_dst(instr_copy, num_reg_dsts)))
+                return false;
+            ++num_reg_dsts;
+        }
+    }
+
+    /* Check that we have the same number of destination operands that are registers.
+     */
+    if (num_reg_dsts != instr_num_dsts(instr_copy))
+        return false;
+
+    /* Check that register source operands are the same.
+     */
+    uint original_num_srcs = (uint)instr_num_srcs(instr_orig);
+    uint num_reg_srcs = 0;
+    for (uint i = 0; i < original_num_srcs; ++i) {
+        opnd_t src_opnd = instr_get_src(instr_orig, i);
+        if (opnd_is_reg(src_opnd)) {
+            // TODO i#6662: we need to compare the virtual register stored in instr_copy
+            // against the converted DR_REG_ (to virtual register) of instr_orig.
+            // Currently we don't have virtual registers, so we can compare the two
+            // DR_REG_ of instr_orig and instr_copy directly.
+            if (!opnd_same(src_opnd, instr_get_src(instr_copy, num_reg_srcs)))
+                return false;
+            ++num_reg_srcs;
+        }
+    }
+
+    /* Check that we have the same number of source operands that are registers.
+     */
+    if (num_reg_srcs != instr_num_srcs(instr_copy))
+        return false;
+
+    return true;
+}
+
+static void
+test_instr_encode_decode_synthetic(void *dc, instr_t *instr, byte *bytes)
+{
+    instr_t *instr_copy = NULL;
+    instr_copy = instr_create(dc);
+    ASSERT(instr_copy != NULL);
+    bool is_isa_mode_set = false;
+    is_isa_mode_set = instr_set_isa_mode(instr_copy, DR_ISA_SYNTHETIC);
+    ASSERT(is_isa_mode_set);
+    is_isa_mode_set = instr_set_isa_mode(instr, DR_ISA_SYNTHETIC);
```

Comment:
Add comment that you are deliberately calling the same function again (o/w looks like it might be a mistake)

Reply:
Done.